src/index.ts

```diff
@@ -279,21 +279,41 @@ class Server {
           data.context = ('vessels.' + app.selfId) as Context
         }
         const now = new Date()
-        data.updates.forEach((update: Update) => {
-          if (typeof update.source !== 'undefined') {
-            update.source.label = providerId
-            if (!update.$source) {
-              update.$source = getSourceId(update.source)
+        data.updates = data.updates
+          .map((update: Partial<Update>) => {
+            if (typeof update.source !== 'undefined') {
+              update.source.label = providerId
+              if (!update.$source) {
+                update.$source = getSourceId(update.source)
+              }
+            } else {
+              if (typeof update.$source === 'undefined') {
+                update.$source = providerId as SourceRef
+              }
             }
-          } else {
-            if (typeof update.$source === 'undefined') {
-              update.$source = providerId as SourceRef
+            if (!update.timestamp || app.config.overrideTimestampWithNow) {
+              update.timestamp = now.toISOString() as Timestamp
             }
-          }
-          if (!update.timestamp || app.config.overrideTimestampWithNow) {
-            update.timestamp = now.toISOString() as Timestamp
-          }
-        })
+
+            if ('values' in update && !Array.isArray(update.values)) {
```

Comment:
How about documenting what this does / why it's here by extracting something like `hasInvalidValues` and ditto for meta?

Reply:
Scratch this - without the `'values'  in update` inline here the values references in the following lines don't work

Reply:
Yeah, I looked at fixing this the other day and ran into the same issue, then I forgot to come back to it.